app/models/transaction.rb

```diff
@@ -21,17 +21,27 @@ class Transaction < ApplicationRecord
 
 	validates :supporter, presence: true
 
+
+	def ordered_payments
+		payments.ordered_query
+	end
+
 	def amount_as_money
     Amount.new(amount||0, nonprofit.currency)
   end
 
-	# def designation
-	# 	donation&.designation
-	# end
-	
-	# def dedication
-	# 	donation&.dedication
-	# end
+	# Handle a completed refund from a legacy Refund object
+	def process_refund(refund)
+		new_refund = subtransaction.process_refund(refund)
+		self.amount = subtransaction.subtransactable.amount
+		transaction_assignments.process_refund(refund)
+		save!
+		publish_updated
+		subtransaction.publish_updated
+		ordered_payments.select{|i| i != new_refund}.each(&:publish_updated)
+		new_refund.publish_created
+		transaction_assignments.first.publish_updated
+	end
```

Comment:
It's a bit hard to understand what's going on here, I would create methods for each part of it to make it clearer or add some comments.

Reply:
Fair point, let me think about how we can do that...

Reply:
I've adjusted this in the latest and I think it makes it clearer.